File: other/MMA2025/H.cpp
#include <bits/stdc++.h>
#include <atcoder/all>
using namespace std;
using namespace atcoder;

#define all(v) v.begin(), v.end()
#define SORT(v) sort(v.begin(), v.end())
#define RSORT(v) sort(v.rbegin(), v.rend())
#define REVERSE(v) reverse(v.begin(), v.end())
#define ll long
#define ld long double

#define int ll

// https://algo-method.com/descriptions/119#h2-61
vector<pair<long long, long long>> prime_factorize(long long N) {
    // 答えを表す可変長配列
    vector<pair<long long, long long>> res;

    // √N まで試し割っていく
    for (long long p = 2; p * p <= N; ++p) {
        // N が p で割り切れないならばスキップ
        if (N % p != 0) {
            continue;
        }

        // N の素因数 p に対する指数を求める
        int e = 0;
        while (N % p == 0) {
            // 指数を 1 増やす
            ++e;

            // N を p で割る
            N /= p;
        }

        // 答えに追加
        res.emplace_back(p, e);
    }

    // 素数が最後に残ることがありうる
    if (N != 1) {
        res.emplace_back(N, 1);
    }
    return res;
}

// https://algo-method.com/descriptions/84#h2-44
// N の約数をすべて求める関数
vector<long long> calc_divisors(long long N) {
    // 答えを表す集合
    vector<long long> res;

    // 各整数 i が N の約数かどうかを調べる
    for (long long i = 1; i * i <= N; ++i) {
        // i が N の約数でない場合はスキップ
        if (N % i != 0) continue;

        // i は約数である
        res.push_back(i);

        // N ÷ i も約数である (重複に注意)
        if (N / i != i) res.push_back(N / i);
    }

    // 約数を小さい順に並び替えて出力
    sort(res.begin(), res.end());
    return res;
}

int pow2(int n) {
    return n * n;
}

int32_t main() {
    int t;
    cin >> t;
    while (t--) {
        int L;
        cin >> L;
        if (L % 2 == 0) {
            int tmp = L / 2;
            auto factors = prime_factorize(tmp);
            int m = factors[0].first * factors[0].second;
            int n = tmp / m;
            if (m * m - n * n < 0) {
                swap(m, n);
            }
            cout << m * m - n * n << " " << 2 * m * n << " " << m * m + n * n << endl;
            assert(pow2(m * m - n * n) + pow2(2 * m * n) == pow2(m * m + n * n));
        } else {
            bool printed = false;
            auto divs = calc_divisors(L);
            for (int i = 0; i < divs.size(); i++) {
                int p = divs[i];
                int q = L / p;
                int gc = gcd(p, q);
                p /= gc;
                q /= gc;
                if (p * p - q * q < 0) {
                    swap(p, q);
                }
                if ((pow2(p) + pow2(q)) % 2 == 0 && (pow2(p) - pow2(q)) % 2 == 0) {
                    cout << (p * p - q * q) / 2 * gc << " " << p * q * gc << " " << (p * p + q * q) / 2 * gc << endl;
                    assert(pow2((p * p - q * q) / 2 * gc) + pow2(p * q * gc) == pow2((p * p + q * q) / 2 * gc));
                    printed = true;
                    break;
                }
            }
            if (!printed) {
                assert(false);
            }
        }
    }
}

File: other/MMA2025/H.py
#!/usr/bin/env python3
import sys
from math import gcd, isqrt

# 元の C++ に合わせた関数群


def prime_factorize(N: int):
    """(p, e) のリストを返す．"""
    res = []
    n = N
    p = 2
    while p * p <= n:
        if n % p != 0:
            p += 1 if p == 2 else 2  # 2 の次は奇数だけ試す小さな最適化
            continue
        e = 0
        while n % p == 0:
            e += 1
            n //= p
        res.append((p, e))
        p += 1 if p == 2 else 2
    if n != 1:
        res.append((n, 1))
    return res


def calc_divisors(N: int):
    """N の全ての約数をソートされたリストで返す．"""
    res = []
    i = 1
    # isqrt を使って整数の平方根まで走査
    r = isqrt(N)
    for i in range(1, r + 1):
        if N % i != 0:
            continue
        res.append(i)
        if N // i != i:
            res.append(N // i)
    res.sort()
    return res


def pow2(n: int):
    return n * n


def main():
    t = int(input())
    out_lines = []
    for _ in range(t):
        L = int(input())
        if L % 2 == 0:
            tmp = L // 2
            factors = prime_factorize(tmp)
            # 元の C++ と同様に，最初の素因子 (prime, exponent) を掛け合わせて m を作る
            if not factors:
                raise ValueError(
                    "tmp == 1 のとき，元の実装は factors[0] を参照します．入力制約を確認してください．"
                )
            m = factors[0][0] * factors[0][1]
            n = tmp // m
            if m * m - n * n < 0:
                m, n = n, m
            a = m * m - n * n
            b = 2 * m * n
            c = m * m + n * n
            # アサートはデバッグ用．必要なら外してください．
            assert pow2(a) + pow2(b) == pow2(c)
            out_lines.append(f"{a} {b} {c}")
        else:
            divs = calc_divisors(L)
            printed = False
            for p in divs:
                q = L // p
                gc = gcd(p, q)
                p0 = p // gc
                q0 = q // gc
                p, q = p0, q0
                if p * p - q * q < 0:
                    p, q = q, p
                if (pow2(p) + pow2(q)) % 2 == 0 and (pow2(p) - pow2(q)) % 2 == 0:
                    a = (p * p - q * q) // 2 * gc
                    b = p * q * gc
                    c = (p * p + q * q) // 2 * gc
                    assert pow2(a) + pow2(b) == pow2(c)
                    out_lines.append(f"{a} {b} {c}")
                    printed = True
                    break
            if not printed:
                assert False
    sys.stdout.write("\n".join(out_lines))

if __name__ == "__main__":
    main()
